MSAL Automation Test Case Repair

fadidurah opened the pull request:
**What:** Fixing broken test cases in the msalautomationapp testpass. Also went through the  testpass and validated that all of the testcases match the ADO Test Plan instructions https://identitydivision.visualstudio.com/Engineering/_testPlans/execute?planId=1905195&suiteId=1905202.

**Why:** Fixing broken test cases and making sure they are all up to date with the test steps in ADO will lower the manual testing burden in future releases. Fixed cases can be used for reference in future automation cases.

**How:** Ran the entire msalautomationapp testpass and went through each case to find any differences or test failures. Ran the tests on the automation pipelines to insure significant increase in reliability and consistent passrates.

**Testing:** Ran the tests frequently on the ui automation pipelines, passrate and consistency has gone up dramatically. Pipelines:
https://dev.azure.com/IdentityDivision/Engineering/_build?definitionId=1743
https://dev.azure.com/IdentityDivision/Engineering/_build?definitionId=1742&_a=summary

fadidurah committed:
first round of fixes, mostly whitespace. Validated that all steps from ADO were followed

fadidurah committed:
Fixed test cases, updated common

fadidurah committed:
Merge branch 'dev' of https://github.com/AzureAD/microsoft-authentication-library-for-android into fadi/repair

fadidurah committed:
Handle chrome first run in 99563

shahzaibj commented on msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/local/TestCase99656.java:
```diff
@@ -40,13 +40,15 @@
 
 // Interactive auth with force_login and step-up MFA
 // https://identitydivision.visualstudio.com/DefaultCollection/IDDP/_workitems/edit/99656
+@Ignore("https://identitydivision.visualstudio.com/Engineering/_workitems/edit/1886086")
```
For this and any other AUTO_MFA test cases, Ryan was saying that we can try clicking on the "Call <Number>" button and see if that let's the flow proceed.

![image](https://user-images.githubusercontent.com/37125644/175165624-0e437d46-0226-4abe-afdc-af61c2bb2f69.png)

Can you try that out? If that works then we can remove the ignore annotation

fadidurah replied:
I tried this, didn't seem to work. the called number doesn't responds and the sign in request doesn't finish. Same thing with text, but that's expected since I need to enter the code myself.

shahzaibj replied:
Probably need to follow up with Ryan on this

fadidurah committed:
update versions, add other fixes

fadidurah committed:
Update common

fadidurah committed:
Attempting some fixes

fadidurah committed:
fixed submodule

fadidurah committed:
update common, minor changes to cases

fadidurah committed:
Adding some test improvements, and updating to take new common changes

fadidurah committed:
Merge branch 'dev' into fadi/repair

fadidurah committed:
remove unused import

fadidurah committed:
Remove sign in audience

fadidurah committed:
Unused Import

fadidurah committed:
Unused Import

fadidurah committed:
update common

fadidurah committed:
Disable google B2C test, Fix 833515

fadidurah committed:
Actually fix 833515

fadidurah committed:
update common

fadidurah committed:
merge common dev

fadidurah committed:
fix common

fadidurah committed:
reset common

fadidurah committed:
unused import

fadidurah committed:
android manifest issue

fadidurah committed:
Adding thread exception catching to 3 tests

fadidurah committed:
new exception catching rule in common

fadidurah committed:
update common

fadidurah committed:
put exception handle rule later

fadidurah committed:
stability improvements

fadidurah committed:
remove latches where not needed

fadidurah committed:
Removed more un-needed latches, update common

fadidurah committed:
Merge branch 'dev' of https://github.com/AzureAD/microsoft-authentication-library-for-android into fadi/repair

fadidurah committed:
Pull shahzaib's pipeline changes

fadidurah committed:
Move AtPop tests to Broker folder

fadidurah committed:
update common, uncaughtexceptionrule back in primary rules

fadidurah committed:
update common

fadidurah committed:
remove browser clearing after tests

fadidurah committed:
1 second wait between acquiretoken and UI handling

fadidurah committed:
return browser clear

fadidurah committed:
remove @after broswer additions

fadidurah committed:
trying fix for 2 test cases

fadidurah committed:
Small tweaks for 99267 and 1420484

fadidurah committed:
Adding test case reliability fixes

fadidurah committed:
Update common

fadidurah committed:
remove chrome sign out check from 833515

fadidurah committed:
retry on fails for two tests

shahzaibj committed:
Merge dev into fadi/repair

fadidurah committed:
Merge branch 'dev' into fadi/repair

fadidurah committed:
update common

fadidurah committed:
Merge branch 'dev' of https://github.com/AzureAD/microsoft-authentication-library-for-android into fadi/repair

fadidurah committed:
fix get usage in AbstractGuestAccountMsalUiTest

fadidurah committed:
fix package name for 1954181 and 1954183, revet get change

fadidurah committed:
re-add get fix

fadidurah committed:
add check for reload to 938368

fadidurah committed:
switch to lab query rather than creating a new user

fadidurah committed:
removing extra wait

fadidurah committed:
update common, some test fixes

fadidurah committed:
extending temp user creation wait time

fadidurah committed:
Creating another Broker pipeline

Creating another pipeline to take some of the test case load off of the first ui automation MSAL W/ broker pipeline [skip ci]

fadidurah committed:
reorganize broker tests into more packages

fadidurah committed:
merge common, update dev

fadidurah committed:
change target of new pipeline to the iteration packages in msalautomationapp/broker

fadidurah committed:
new packages for tests, created packages for first and second iterations

fadidurah committed:
remove i1, i2

fadidurah committed:
split cases based on package

fadidurah committed:
merge new packages for msal with broker cases

fadidurah committed:
Merge branch 'fadi/extra-msal-broker-pipeline' of https://github.com/AzureAD/microsoft-authentication-library-for-android into fadi/repair

fadidurah committed:
remove second pipeline, seems like something broke

fadidurah committed:
fix detached common

fadidurah committed:
Repackage automation tests

fadidurah committed:
fix 833515

fadidurah committed:
rename api and endshift, move a few more tests

fadidurah committed:
Merge branch 'fadi/test-iteration' of https://github.com/AzureAD/microsoft-authentication-library-for-android into fadi/repair

fadidurah committed:
update pipeline to include new package names

shahzaibj commented on azure-pipelines/ui-automation/msal-broker-BrokerHost-nightly.yml:
```diff
@@ -183,7 +183,7 @@ stages:
                   --environment-variables "clearPackageData=true" \
                   --results-history-name "$(resultsHistoryName)" \
-                  --test-targets "package com.microsoft.identity.client.msal.automationapp.testpass.broker"
+                  --test-targets "package com.microsoft.identity.client.msal.automationapp.testpass.broker.brokerapi, package com.microsoft.identity.client.msal.automationapp.testpass.broker.atpop, package com.microsoft.identity.client.msal.automationapp.testpass.broker.crosscloud, package com.microsoft.identity.client.msal.automationapp.testpass.broker.flw"
```
Is this actually needed? I think just having `com.microsoft.identity.client.msal.automationapp.testpass.broker` should've been good enough because the rest are just sub-packages.

If that is really not working then you may want to try:
`com.microsoft.identity.client.msal.automationapp.testpass.broker.*` and see if that works

fadidurah replied:
This was added so that we don't attempt to run every test case and run into the 45m limit. The packages included here contain about half the test cases. I ended up adding a parameter to feed this through.

shahzaibj commented on msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/crosscloud/TestCase1592465.java:
```diff
@@ -117,7 +122,17 @@ public void test_acquire_token_from_cross_cloud_after_acquiring_token_from_home_
         // Acquire token interactively from cross cloud, expected to get a new access token
-        final OnInteractionRequired crossCloudInteractionHandler = () -> {};
+        final OnInteractionRequired crossCloudInteractionHandler = () -> {
+            final PromptHandlerParameters promptHandlerParameters =
+                    PromptHandlerParameters.builder()
+                            .prompt(PromptParameter.SELECT_ACCOUNT)
+                            .loginHint(userName)
+                            .broker(mBroker)
+                            .staySignedInPageExpected(GlobalConstants.IS_STAY_SIGN_IN_PAGE_EXPECTED)
+                            .build();
+            final AadPromptHandler promptHandler = new AadPromptHandler(promptHandlerParameters);
+            promptHandler.handlePrompt(userName, password);
```
So this was previously passing without a prompt handler I guess.

shahzaibj replied:
I think the Expected Results in the Test Case in ADO are not clear. @iamgusain Can you review that test case in ADO ensure that Expected Results column clearly indicates whether to expect a prompt or not?

fadidurah replied:
This case and `1592510` are behaving differently on pipeline, the second request is not prompting for credentials. On my physical device, the second request always prompts for credentials. I'm probably just going to ignore these two for now and look at them later.

shahzaibj commented on msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/flw/TestCase833515.java:
```diff
@@ -104,57 +119,43 @@ public void test_833515() throws MsalException, InterruptedException, LabApiExce
-        final UiObject passwordField = UiAutomatorUtils.obtainUiObjectWithResourceId("i0118");
-        Assert.assertTrue(passwordField.exists());
+        // can sometimes take a while to actually be signed out
+        Thread.sleep(TimeUnit.SECONDS.toMillis(8));
```
I believe the timeout was for when we were asserting logout in the browser. Since we removed that assertion, do we still need the sleep here?

shahzaibj commented on msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/usgov/arlington/TestCase938368.java:
```diff
@@ -42,10 +48,13 @@
 // Interactive token acquisition with instance_aware=true, login hint present, and federated account,
 // and WW common authority
 // https://identitydivision.visualstudio.com/Engineering/_workitems/edit/938368
+// Adding a retry on failure, sometimes arlington login page fails to load
```
Does it give an error when it fails to load? How often does this happen? 1 in 5 times? 1 in 10 times? 1 in 20 times? or some other frequency?

fadidurah replied:
Sometimes gets the "This page isn't working" error. I'd say around 1 in 15 times it would happen. I also added code in this to refresh the page if the error shows up.

shahzaibj commented on msalautomationapp/src/main/java/com/microsoft/identity/client/msal/automationapp/sdk/MsalSdk.java:
```diff
@@ -101,11 +103,17 @@ public MsalAuthResult acquireTokenInteractive(@NonNull MsalAuthTestParams authTe
 
         pca.acquireToken(acquireTokenParameters);
 
+        // Adding small sleep to make sure acquireToken is up before we start handling UI
```
What happens otherwise?

fadidurah replied:
I think I did this in response to some cases having a UiSelector failure where it wouldn't be able to locate the username or password box.

shahzaibj reviewed (approved):
Minor comments, otherwise lgtm

somalaya commented on msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/atpop/TestCase1954183.java:
```diff
@@ -20,7 +20,7 @@
 //  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 //  THE SOFTWARE.
-package com.microsoft.identity.client.msal.automationapp.testpass.atpop;
+package com.microsoft.identity.client.msal.automationapp.testpass.broker.atpop;
```
Why is this file moved under broker folder?

fadidurah replied:
I think when I did some early repackaging, I thought all of these test were brokered so I moved them to a folder in broker, I can move 1954181 and 1954183 to msalonly and give them a package in that.

somalaya commented on msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/crosscloud/TestCase1592510.java:
```diff
@@ -117,7 +121,16 @@ public void test_acquire_token_from_home_cloud_after_acquiring_token_from_cross_
+        final OnInteractionRequired homeCloudInteractionHandler = () -> {
+            final PromptHandlerParameters promptHandlerParameters =
+                    PromptHandlerParameters.builder()
+                    .prompt(PromptParameter.SELECT_ACCOUNT)
```
Question : When we expect to see an interactive prompt - Are we also expecting to see a password prompt? Or just seeing the username field is enough in this case? Because setting the prompt type as SELECT_ACCOUNT will not autofill the password field but only autofills the username field.

fadidurah replied:
This test case is only showing the password prompt, with SELECT_ACCOUNT that prompt is being autofilled. The expected behavior for this case and 1592465 is a bit unclear, so i'm ignoring them for now.

fadidurah committed:
Merge pull request #1669 from AzureAD/rapong/repackageAutomationTests

Repackage automation tests

fadidurah committed:
update pipeline YAMLs

fadidurah committed:
add parameter for packages to pipeline

fadidurah committed:
small typo in yml

fadidurah committed:
yaml syntax issue

fadidurah committed:
fix yaml parameter

fadidurah committed:
fix yaml parameter

fadidurah committed:
ignore 1592465, 1592510

fadidurah committed:
address some comments, move brokerless atPop tests to msalonly.

fadidurah committed:
Include b2c in MSAL w/o broker Automation Pipeline

rpdome committed:
Move MSAL Tests, bring back some tests too

fadidurah committed:
Merge pull request #1672 from AzureAD/rapong/moveMSALTests

Move MSAL Tests, bring back some tests too (Dome)

fadidurah committed:
Ignore TestCase1592510, expected behavior unclear
